### PR TITLE
fix: aligns lifecycle calls in JUnit 5 with JUnit 4 behavior

### LINKED ISSUE #309: 1.7.0.Alpha7 does not call JUnit lifecycle hooks?
I maintained a [jakartaee9 starter template](https://github.com/hantsy/jakartaee9-starter-boilerplate) which is provided to kickstart a Jakarta EE 9 project.

I am not sure what is changed in the latest 1.7.0.Alpha7 in today's PR created by Renovate.

From 1.7.0.Alpha6->1.7.0.Alpha7, the following simple test failed due to a NullPointerException. 

```java
@ExtendWith(ArquillianExtension.class)
public class GreetingResourceTest {
    private final static Logger LOGGER = Logger.getLogger(GreetingResourceTest.class.getName());
    
    @Deployment(testable = false)
    public static WebArchive createDeployment() {
        return ShrinkWrap.create(WebArchive.class)
                .addClass(GreetingMessage.class)
                .addClass(GreetingService.class)
                .addClasses(GreetingResource.class, JaxrsActivator.class)
                // Enable CDI (Optional since Java EE 7.0)
                .addAsWebInfResource(EmptyAsset.INSTANCE, "beans.xml");
    }
    
    @ArquillianResource
    private URL base;
    
    private Client client;
    
    @BeforeEach
    public void setup() {
        LOGGER.info("call BeforeEach");
        this.client = ClientBuilder.newClient();
        //removed the Jackson json provider registry, due to OpenLiberty 21.0.0.1 switched to use Resteasy.
    }
    
    @AfterEach
    public void teardown() {
        LOGGER.info("call AfterEach");
        if (this.client != null) {
            this.client.close();
        }
    }
    
    @Test
    @DisplayName("Given a name:`JakartaEE` should return `Say Hello to JakartaEE`")
    public void should_create_greeting() throws MalformedURLException {
        LOGGER.log(Level.INFO, " client: {0}, baseURL: {1}", new Object[]{client, base});
        final WebTarget greetingTarget = this.client.target(new URL(this.base, "api/greeting/JakartaEE").toExternalForm());
        try (final Response greetingGetResponse = greetingTarget.request()
                .accept(MediaType.APPLICATION_JSON)
                .get()) {
            assertThat(greetingGetResponse.getStatus()).isEqualTo(200);
            assertThat(greetingGetResponse.readEntity(GreetingMessage.class).getMessage()).startsWith("Say Hello to JakartaEE");
        }
    }
}
```
The client is **null** there in the test method `should_create_greeting`.

Check the build result from Github actions(Ubuntu, Java 8).

https://github.com/hantsy/jakartaee9-starter-boilerplate/pull/99

### PR DESCRIPTION
**Pull request by lprimak**
- @BeforeAll / @AfterAll callbacks are always called on the client, consistent with JUnit 4 behavior
- restored calls to @BeforeEach / @AfterEach in client mode, were only being called on the server

#### Short description of what this resolves:
Fixes problems encountered in #288 and #301
Fixes #309 

## Note: Use "ignore whitespace changes" to review

### CODE REVIEW

**Comment by lprimak** [comment]
@bartoszmajsak Can you review this please?
Thank you!

**Comment by bartoszmajsak** [comment]
Thanks @lprimak! Merging #301 was giving me a bit of mixed feeling seeing this without any tests, but here we are :) I wonder though - do you think we could add some of the reproducers you created (and maybe more tests as we evolve) to be part of repo/build? This way we can avoid such challenges going forward.

File: junit5/core/src/main/java/org/jboss/arquillian/junit5/ArquillianExtension.java [hunk 1]
```diff
@@ -95,44 +92,44 @@ public void interceptTestMethod(Invocation<Void> invocation, ReflectiveInvocatio
...
+
     @Override
     public void interceptBeforeAllMethod(Invocation<Void> invocation,
       ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
       if (isInsideArquillian.test(extensionContext)) {
-        invocation.proceed();
-      } else {
         invocation.skip();
+      } else {
+        invocation.proceed();
```
Comment (bartoszmajsak) [inline junit5/core/src/main/java/org/jboss/arquillian/junit5/ArquillianExtension.java:122]: Why are we reversing the logic here? We don't want to have `@BeforeAll` executed as Arquillian test?

Comment (dmatej) [reply junit5/core/src/main/java/org/jboss/arquillian/junit5/ArquillianExtension.java:122]: Seems to me we are going backwards  to the same issue we had two years back - it depends. It makes sense for both use cases, sometimes.

I think we need another annotation for this and decide what should be the default - run on the client side or on the server side? And warn before undefined order if both variants combined (multiple BeforeEach annotated methods). 

Or am I wrong?

Comment (lprimak) [reply junit5/core/src/main/java/org/jboss/arquillian/junit5/ArquillianExtension.java:122]: Not the same case David :). This is the simplest case

Comment (lprimak) [reply junit5/core/src/main/java/org/jboss/arquillian/junit5/ArquillianExtension.java:122]: @bartoszmajsak the logic is reversed to accomplish the fix for the issue. These methods have to be only executed on the client and not in container as they are executed that way in JUnit 4

I’ll try to add the test for this as well

Comment (lprimak) [reply junit5/core/src/main/java/org/jboss/arquillian/junit5/ArquillianExtension.java:122]: I have added the tests. However, they are excluded for now. This requires integration test setup with app server as far as I am aware, and it's not done in Arquillian just yet, so I am not sure how to introduce that kind of tests.
Saying that, it's there and works when enabled :)

**Comment by lprimak** [comment]
@bartoszmajsak added tests

**Comment by lprimak** [comment]
> Thanks @lprimak! Merging #301 was giving me a bit of mixed feeling seeing this without any tests, but here we are :) I 

Well, the change was so simple :) And also if there were no tests originally for client invocations, they would not have helped :) Also it's not at all clear that the behavior in JUnit4 container is correct as @dm says.

However, with this PR, status quo is maintained and nothing is broken, so there's that...

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
Few minor points I wanted to raise with you before we can move this forward.

P.S. I've been away from the keyboard the whole last week hence the silence.

File: junit5/core/pom.xml [hunk 2]
```diff
@@ -1,42 +1,95 @@
...
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>fish.payara.arquillian</groupId>
+            <artifactId>arquillian-payara-server-remote</artifactId>
+            <version>2.3.3</version>
+            <scope>test</scope>
+        </dependency>-->
+
```
[the remark this answers is not in the document]
Comment (lprimak) [reply junit5/core/pom.xml:61]: I will put the new tests into another PR so you can decide what to do with them, and remove from this PR.
@kifj has ported tests so now this code is tested - thank you!
Above can move this PR forward

File: junit5/core/pom.xml [hunk 3]
```diff
@@ -1,42 +1,95 @@
...
-      <artifactId>junit-jupiter-api</artifactId>
-      <scope>provided</scope>
-    </dependency>
-    <dependency>
-      <groupId>org.junit.platform</groupId>
-      <artifactId>junit-platform-engine</artifactId>
-      <scope>provided</scope>
-    </dependency>
-  </dependencies>
+      <modelVersion>4.0.0</modelVersion>
```
[the remark this answers is not in the document]
Comment (lprimak) [reply junit5/core/pom.xml:3]: will be removed, no need

File: junit5/core/pom.xml [hunk 4]
```diff
@@ -1,42 +1,95 @@
...
+        </dependency>-->
+
+        <dependency>
+            <groupId>org.jboss.shrinkwrap</groupId>
+            <artifactId>shrinkwrap-api</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.jboss.arquillian.container</groupId>
+            <artifactId>arquillian-container-test-api</artifactId>
```
[the remark this answers is not in the document]
Comment (lprimak) [reply junit5/core/pom.xml:69]: I will add another PR with the new tests

File: junit5/core/src/test/java/org/jboss/arquillian/junit5/lifecycle/CreateFileTestCase.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import javax.naming.NamingException;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import org.junit.jupiter.api.Test;
+
+/**
+ *
+ * @author lprimak
+ */
+public class CreateFileTestCase {
+    static final String TMP_FILE_NAME ="/tmp/arqBeforeTest-5123";
```
[the remark this answers is not in the document]
Comment (lprimak) [reply junit5/core/src/test/java/org/jboss/arquillian/junit5/lifecycle/CreateFileTestCase.java:31]: DIdn't know that existed, I will try it

[the remark this answers is not in the document]
Comment (lprimak) [reply junit5/core/src/test/java/org/jboss/arquillian/junit5/lifecycle/CreateFileTestCase.java:31]: `@TempDir` gets deleted by JUnit before all the tests are completed. However, it's useful to get the target directory

[the remark this answers is not in the document]
Comment (lprimak) [reply junit5/core/src/test/java/org/jboss/arquillian/junit5/lifecycle/CreateFileTestCase.java:31]: Well, this is turning out quite the task! `@TempDir` returns different directory for every unit test, which is not what I want. Working around it is turning out to be a chore!

**Comment by lprimak** [comment]
@bartoszmajsak ready to go!

File: junit5/core/src/test/java/org/jboss/arquillian/junit5/lifecycle/CreateFileTestCase.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply junit5/core/src/test/java/org/jboss/arquillian/junit5/lifecycle/CreateFileTestCase.java]: Sure, and this way you can run tests in parallel (or even builds). I guess you could read this variable during the test execution so you should have a knowledge of the actual directory during the test run, right? But lets see in the coming PR. Thanks a lot for all your hard work!

**Review by bartoszmajsak — APPROVED** [review]

**Pull request merged**